test/widgets/subscription_list_test.dart

```diff
@@ -0,0 +1,146 @@
+
+    await tester.pumpWidget(
+      MaterialApp(
+        home: GlobalStoreWidget(
+          child: PerAccountStoreWidget(
+            accountId: eg.selfAccount.id,
+            child: const SubscriptionListPage()))));
+
+    // global store, per-account store
+    await tester.pumpAndSettle();
+  }
+
+  bool isPinnedHeaderInTree() {
+    return find.text('Pinned').evaluate().isNotEmpty;
+  }
+
+  bool isUnpinnedHeaderInTree() {
+    return find.text('Unpinned').evaluate().isNotEmpty;
+  }
+
+  int getItemCount() {
+    return find.byType(SubscriptionItem).evaluate().length;
+  }
+
+  testWidgets('smoke', (tester) async {
+    await setupStreamListPage(tester, subscriptions: []);
+    check(getItemCount()).equals(0);
+    check(isPinnedHeaderInTree()).isFalse();
+    check(isUnpinnedHeaderInTree()).isFalse();
+  });
+
+  testWidgets('basic subscriptions', (tester) async {
+    await setupStreamListPage(tester, subscriptions: [
+      eg.subscription(eg.stream(streamId: 1), pinToTop: true),
+      eg.subscription(eg.stream(streamId: 2), pinToTop: true),
+      eg.subscription(eg.stream(streamId: 3), pinToTop: false),
+    ]);
+    check(getItemCount()).equals(3);
+    check(isPinnedHeaderInTree()).isTrue();
+    check(isUnpinnedHeaderInTree()).isTrue();
+  });
+
+  testWidgets('only pinned subscriptions', (tester) async {
+    await setupStreamListPage(tester, subscriptions: [
+      eg.subscription(eg.stream(streamId: 1), pinToTop: true),
+      eg.subscription(eg.stream(streamId: 2), pinToTop: true),
+    ]);
+    check(getItemCount()).equals(2);
+    check(isPinnedHeaderInTree()).isTrue();
+    check(isUnpinnedHeaderInTree()).isFalse();
+  });
+
+  testWidgets('only unpinned subscriptions', (tester) async {
+    await setupStreamListPage(tester, subscriptions: [
+      eg.subscription(eg.stream(streamId: 1), pinToTop: false),
+      eg.subscription(eg.stream(streamId: 2), pinToTop: false),
+    ]);
+    check(getItemCount()).equals(2);
+    check(isPinnedHeaderInTree()).isFalse();
+    check(isUnpinnedHeaderInTree()).isTrue();
+  });
+
+  testWidgets('subscription sort', (tester) async {
+    await setupStreamListPage(tester, subscriptions: [
+      eg.subscription(eg.stream(streamId: 1, name: 'd'), pinToTop: true),
+      eg.subscription(eg.stream(streamId: 2, name: 'c'), pinToTop: false),
+      eg.subscription(eg.stream(streamId: 3, name: 'b'), pinToTop: true),
+      eg.subscription(eg.stream(streamId: 4, name: 'a'), pinToTop: false),
+    ]);
+    check(isPinnedHeaderInTree()).isTrue();
+    check(isUnpinnedHeaderInTree()).isTrue();
+
+    final streamListItems = tester.widgetList<SubscriptionItem>(find.byType(SubscriptionItem)).toList();
+    check(streamListItems.length).equals(4);
+    check(streamListItems[0].subscription.streamId).equals(3);
+    check(streamListItems[1].subscription.streamId).equals(1);
+    check(streamListItems[2].subscription.streamId).equals(4);
+    check(streamListItems[3].subscription.streamId).equals(2);
```

Comment:
Similarly here: imagine writing a change that turns out to break this test. This is basically the moment where the test fulfills its purpose — the point of a test is that if we write a change that breaks it, the test tells us there's something wrong with our change, or at least something that contradicts a previous expectation and we need to revise the expectation. So the value of the test is tied to how well it tells us what it is that's wrong or unexpected about our change.

Here, that mission will be much better accomplished if the test failure message tells us about the full list of these streams, rather than just one — the full list is how we can see how the sorting ended up and how that differs from what was expected.

So let's get the observed result into the form of a list that we can compare with `deepEquals`. Could be like `check(streamListItems.map(…)).deepEquals([3, 1, 4, 2])`.